### DISCUSSION TITLE
FastAPI custom header versioning with requests under duplicate paths.

### DISCUSSION THREAD

**Discussion by Bartket** [fastapi/fastapi#8784] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Header, Request, HTTPException

app = FastAPI()

@app.get("/greet")
def greet_version(request: Request, X_API_Version: int = Header(int(1))):
    """
    This request uses X-API-Version header to determine version

    - **X-API-Version**: request must have a version either 1 or 2 is accepted.
    """
    version = request.headers.get('X-API-Version')
    if version == '1': 
        return {'message': "Hello"}
    if version == '2': 
        return {'notice': "Hello"}
    else: 
        raise HTTPException(status_code=404, detail="Version not found")
```

### Description

I want to create FastAPI app with custom header versioning and the only solution I found is to add the check in the request itself what is the header value and based on that return the adjusted response. This approach is not very good in terms of code organization(nested if-else for versioning), data validation ex. what in case version two would require different query parameters, I cannot set both in the same request. Also, this will not be described properly with OpenAPI docs. I would rather have requests under the same path and define them according to version changes based on additional header parameters but I am not sure how to approach this especially since two requests cannot have the same path and I don't want to use redirects ex. with the decorator:
```python
@app.get("/greet")
@header_version(1)
def greet_version(request: Request, X_API_Version: int = Header(int(1))):
        return {'message': "Hello"}

@app.get("/greet")
@header_version(2)
def greet_version(request: Request, X_API_Version: int = Header(int(2))):
        return {'notice': "Hello"}
```

### Operating System

Linux

### FastAPI Version

0.66.0

### Python Version

3.8.10

**Comment by panla** [comment]
dependencies <https://fastapi.tiangolo.com/tutorial/dependencies/>

**Comment by Bartket** [comment]
@panla I know I can add (for example) a global dependency to check for versions, I wonder how I could have the same path operation multiple times based on the header value condition. With the current option, I cannot really document changes with OpenAPI as I could have with path-based versioning.

For now, I have something like this:

```python
from fastapi import FastAPI, Header, Request, Depends
from enum import Enum

class VersionName(str, Enum):
    V1 = "V1"
    V2 = "V2"

async def verify_version(X_API_Version: VersionName = Header(...)):
    return X_API_Version

app = FastAPI(dependencies=[Depends(verify_version)])

@app.get("/items")
async def read_items(request: Request):
    version = request.headers.get('X-API-Version')
    if version == VersionName.V1:
        return {"item_id": "Foo"}
    if version == VersionName.V2:
        return {"item_id": "Foo2"}
```

But I would prefer to have something like this with the difference being a custom header check:
```python
@app.get("/items")
async def read_itemsv1():
    return {"item_id": "Foo"}

@app.get("/items")
async def read_itemsv2():
    return {"item_id": "Foo2"}
```
Normally FastAPI will only register only one path operation unless it uses a different method (GET/POST/DELETE) etc. 

Could you elaborate on how to use dependencies for that, or you meant a solution similar to my first proposition?

**Comment by HHK1** [comment, marked as answer]
You could do something very similar to this: https://github.com/tiangolo/fastapi/issues/3910#issuecomment-1024132161

You'll need to adapt the middleware to read your custom header, but then the rest is pretty similar
